### DISCUSSION TITLE
Textual CLI with new python version not updating

### DISCUSSION THREAD

**Discussion by arjunj132** [Textualize/textual#4359] [Q&A] [answered]
Earlier today, I install python 3.12 using `asdf` (on my mac) and its working fine with the command `python3` and `pip3`, with the updated version, but not with the `textual` command.

So when I downloaded it, all my dependencies were erased. Just running my textual app using `python3 <app>.py` doesn't work. But when I use `textual run <app>.py`, it works fine. And downloading a dependency using `pip` and using it in my app will result in a error saying that dependency is not found, meaning that `textual run ...` is using the older python. 

I kind of bypassed this using

```bash
sudo nano ~/.zshrc
```

And then adding an alias:

```bash
alias textual="python3 -m textual_dev"
```

(The `python3 -m textual_dev` uses the new python version and works)

My older python is still accessible using `/usr/bin/python3`, like the system executable, so I'm guessing the textual cli is using that version.

```bash
____@macbook % python3 --version
Python 3.12.2
____@macbook % /usr/bin/python3 --version
Python 3.9.6
```

I'm proposing just a textual cli option to set the python executable to like `python3` or `py` or `python` or `/usr/bin/python3` or whatever, or at least a flag for it.

**Comment by TomJGooding** [comment]
Sorry but I'm struggling to understand the issue here. Are you suggesting you think there's a Textual problem with Python 3.12? Or is this a problem with multiple Python versions installed, which has nothing to do with Textual?

**Reply by arjunj132** [reply]
Yeah @TomJGooding, I wasn't very clear with the problem. So the issue is that when I updated, the `textual` command still uses my older python, which shouldn't necessarily be a problem, but it runs my project using my old python, which is now a problem. Im proposing a flag, like `--exec python3` to set an executable manually.

**Reply by arjunj132** [reply]
Its more like a feature request than a bug, when I think about it.

**Reply by davep** [reply]
It would be normal to have `textual-dev` installed along with `textual` in the same virtual environment, which should mean this isn't an issue while you're using the `textual` command during development.

**Reply by arjunj132** [reply]
@davep Yes, you're correct, in a virtual environment this shouldn't be a problem. But without one, and if I don't uninstall textual-dev, and I change my default python version, and keep the old python, textual-dev will still use the older version, and will run my project in an older version. So just a flag to set a python executable when running my project inside the textual command, like, `textual run app.py --exec python3` would be nice.

**Reply by davep** [reply]
What would be the purpose of installing the devtools in something other than your virtual environment that you're developing within?

**Reply by arjunj132** [reply]
I installed the devtools globally. Should I have installed it in my virtual environment?

**Reply by arjunj132** [reply]
Then its more of a me problem, sorry for wasting your time...

**Reply by davep** [reply, marked as answer]
I would always install the devtools (and Textual) into my virtual environment.

And don't worry, you're not wasting anyone's time with questions. :-)

**Reply by arjunj132** [reply]
Thank you so much @davep!
